chain/chain/src/chain.rs

```diff
@@ -531,20 +528,115 @@ impl Chain {
+    // 3. `clear_data()` contains of four parts:
+    //    a. Block Reference Map creating, if it not exists
+    //    b. Define clearing height as `height`; the highest height for which we want to clear the data
+    //    c. Forks Clearing at height `height` down to tail.
+    //    d. Canonical Chain Clearing from tail up to height `height`
+    // 4. Before actual clearing is started, Block Reference Map should be built.
+    // 5. It's recommended to execute `clear_data()` every time when block at new height is added.
+    //
+    // Forks Clearing:
+    // 1. Any fork which ends up on height `height` INCLUSIVELY and earlier will be completely deleted
+    //    from the Store with all its ancestors up to the ancestor block where fork is happened
+    //    EXCLUDING the ancestor block where fork is happened.
+    // 2. The oldest ancestor block always remains on the Canonical Chain by property 2.
+    // 3. All forks which end up on height `height + 1` and further are protected from deletion and
+    //    no their ancestor will be deleted (even with lowest heights).
+    // 4. `clear_forks_data()` handles forks clearing for fixed height `height`.
+    //
+    // Canonical Chain Clearing:
+    // 1. Blocks on the Canonical Chain with the only descendant (if no forks started from them)
+    //    are unlocked for Canonical Chain Clearing.
+    // 2. If Forks Clearing ended up on the Canonical Chain, the block may be unlocked
+    //    for the Canonical Chain Clearing. There is no other reason to unlock the block exists.
+    // 3. All the unlocked blocks will be completely deleted
+    //    from the tail up to the height `height` EXCLUSIVELY.
+    // 4. (Property 3, GC invariant). There is always only one block with the lowest height (tail)
+    //    (based on property 1) and it's always on the Canonical Chain (based on property 2).
+    //
+    // Example:
+    //
+    // height: 101   102   103   104
+    // --------[A]---[B]---[C]---[D]
+    //          \     \
+    //           \     \---[E]
+    //            \
+    //             \-[F]---[G]
+    //
+    // 1. Let's define clearing height = 102. It this case fork A-F-G is protected from deletion
+    //    because of G which is on height 103. Nothing will be deleted.
+    // 2. Let's define clearing height = 103. It this case Fork Clearing will be executed for A
+    //    to delete blocks G and F, then Fork Clearing will be executed for B to delete block E.
+    //    Then Canonical Chain Clearing will delete blocks A and B as unlocked.
+    //    Block C is the only block of height 103 remains on the Canonical Chain (invariant).
+    //
+    pub fn clear_data(&mut self, trie: Arc<Trie>) -> Result<(), Error> {
         let mut chain_store_update = self.store.store_update();
         let head = chain_store_update.head()?;
+        let tail = chain_store_update.tail()?;
         let height_diff = NUM_EPOCHS_TO_KEEP_STORE_DATA * self.epoch_length;
```

Comment:
Didn't get clearly what's dangerous with long epoch without blocks.
In case of having no blocks in two epochs in a row `process_block` will fail. This happens strictly earlier than GC comes (it needs 5 epochs for now).

Reply:
>having no blocks in two epochs in a row

Epochs don't start at every `epoch_length` height.
If many validators are offline and blocks don't get finalized, we don't start a new epoch. In this scenario we shouldn't garbage collect because we want to be able to switch to a different fork.
If the set of blocks to gc is defined in terms of epochs, we shouldn't be looking at heights only.

Reply:
Misha's concern is not long epochs without blocks, but rather long epochs WITH blocks. The epoch doesn't switch unless it has a final block, so if a finality gadget stops finalizing blocks, the epoch will have blocks, but will not end.

Reply:
How does it affect? We use `NUM_EPOCHS_TO_KEEP_STORE_DATA` which is constant and self.epoch_length which is config thing.

Reply:
If affects us because a single epoch can be longer that `NUM_EPOCHS_TO_KEEP_STORE_DATA` x `epoch_length`.

The actual epoch length can deviate from the desired.
However in practice it can only _significantly_ deviate if more than 1/3 of participants are offline for more than several days.

Reply:
I see. It will break State Syncing completely.
For very long epochs we will start State Syncing from very early point then Block Syncing will try to download too many blocks.
1. Syncing will take too much time.
2. Everything will die if epoch length is longer than amount of memory needed to store all the data.
Do we have any protection from this scenario?

Reply:
@Kouprin doesn't that mean we have a memory leak somewhere? Most data should be stored on disk.

Reply:
@bowenwang1996 According to impossibility of getting OOM, does this page make sense? https://www.notion.so/nearprotocol/20-tests-with-various-adversarial-behaviors-810780e8a3054ec9b132000888e777a8?p=3645cdc63e254a048f00c994ac495224

Reply:
It does make sense. I am not saying that it is currently impossible to have memory leaks. Rather I am saying that we should not have memory leaks. If we find any we should fix them.

Reply:
Am I right, that each case which leads us to OOM will be fixed and there is no reasonable case of handling OOM?

Reply:
I think so but I cannot vouch for it :)

Reply:
> there is no reasonable case of handling OOM?

Yes, applications have no way to handle OOM event, since they are killed with SIGKILL, and there is no way to handle SIGKILL. If we happen to need to reduce the OOM involvement, we should implement a watchdog checking the amount of free memory and active proactively (sometimes it is better to shut down gracefully, than do recovery from an inconsistent state).

Reply:
Regarding this PR, I'm trying to figure out is there an issue with non-finalized epochs or not. If yes, we should work on finalizing epochs instead of fixing GC at this point. If no, we need to have strong confirmation why.

OOM questions are related to figuring out possibility of the issue described above.